src/util/ringdelaybuffer.cpp

```diff
@@ -0,0 +1,215 @@
+* WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+*/
+
+/*
+* The text above constitutes the entire PortAudio license; however,
+* the PortAudio community also makes the following non-binding requests:
+*
+* Any person wishing to distribute modifications to the Software is
+* requested to send the modifications to the original developer so that
+* they can be incorporated into the canonical version. It is also
+* requested that these non-binding requests be included along with the
+* license above.
+*/
+
+#include "ringdelaybuffer.h"
+
+RingDelayBuffer::RingDelayBuffer(SINT bufferSize)
+        : m_bufferSize(bufferSize),
+          m_readPos(0),
+          m_writePos(0),
+          m_ringFullMask((bufferSize * 2) - 1),
+          m_ringMask(bufferSize - 1),
+          m_jumpLeftAroundMask(0),
+          m_buffer(bufferSize) {
+    // TODO(davidchocholaty) Handle to allow only power of two for the size of the ring delay buffer.
+
+    // Set the ring delay buffer items to 0.
+    CSAMPLE* bufferData = m_buffer.data();
+    memset(bufferData, 0, sizeof(*bufferData) * bufferSize);
+}
+
+SINT RingDelayBuffer::read(CSAMPLE* pBuffer, SINT itemsToRead) {
+    const SINT available = getReadAvailable();
+
+    VERIFY_OR_DEBUG_ASSERT(itemsToRead <= available) {
+        itemsToRead = available;
+    }
+
+    const SINT position = m_readPos & m_ringMask;
+
+    // Check to see if the read is not contiguous.
+    if ((position + itemsToRead) > m_bufferSize) {
+        // Read is not contiguous.
+        SINT firstDataBlockSize = m_bufferSize - position;
+
+        memcpy(pBuffer, m_buffer.data(position), firstDataBlockSize * sizeof(CSAMPLE));
+        pBuffer = pBuffer + firstDataBlockSize;
+
+        // The second data part is the start of the ring buffer.
+        memcpy(pBuffer, m_buffer.data(), (itemsToRead - firstDataBlockSize) * sizeof(CSAMPLE));
+    } else {
+        // Read is contiguous.
+        memcpy(pBuffer, m_buffer.data(position), itemsToRead * sizeof(CSAMPLE));
+    }
+
+    // Calculate the new read position. If the new read position
+    // is after the ring delay buffer end, move it around from the start
+    // of the ring delay buffer. If the extra bit for the read position
+    // is set as empty and the jump will cross the right side of the buffer,
+    // the m_ringFullMask has to be XORed with the m_jumpLeftAroundMask,
+    // to handle the situation, when the read position extra bit can't be set
+    // as full for jumping to left. Otherwise, use only the m_ringFullMask
+    // for masking.
+    if ((position > m_bufferSize) && (m_readPos < m_bufferSize)) {
+        // position > m_bufferSize  -> cross the right side of the buffer
+        // m_readPos < m_bufferSize -> extra bit is set as empty
+        m_readPos = (m_readPos + itemsToRead) & (m_ringFullMask ^ m_jumpLeftAroundMask);
+        m_jumpLeftAroundMask = 0;
+    } else {
+        m_readPos = (m_readPos + itemsToRead) & m_ringFullMask;
+    }
```

Comment:
I'm severely struggling to understand this. 
Why can't we just do `m_readPos = (position + itemsToRead) % m_bufferSize`?
That removes the power-of-two size limitation, make the bit-twiddling stuff unnecessary, etc.
We're not calculating the integer mod for each sample, but only once per write call, we can afford that.

Reply:
The complexity introduced by the current scheme makes it very, very hard for me to review the rest of the code, so I'll refrain from that until we can clear this up.

Reply:
To your advice that the modulo operation can be used instead of the bitwise operation. If we do not allow the reading of uninitialized values, it may be used.